### PR TITLE
Old optional

### PR DESCRIPTION
**Pull request by drewdzzz**
The patch populates `mpp::encode` and `mpp::decode` functions with optional objects support. An empty optional is encoded as `MP_NIL`, and `MP_NIL` is decoded as an empty optional. Otherwise, underlying objects are encoded/decoded just like without optional.

### CODE REVIEW

File: src/mpp/Dec.hpp [hunk 1]
```diff
@@ -82,6 +82,10 @@ constexpr auto detectFamily()
 		      "Can't decode to constant type");
 	if constexpr (is_wrapped_family_v<T>) {
 		return family_sequence<T::family>{};
+	} else if constexpr (tnt::is_optional_v<U>) {
+		return family_sequence_populate<compact::MP_NIL>(detectFamily<tnt::value_type_t<U>>());
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:86]: Nit: what's the advantage over

```suggestion
		return family_sequence<compact::MP_NIL, detectFamily<tnt::value_type_t<U>>()>{};
```
?

File: src/mpp/Dec.hpp [hunk 2]
```diff
@@ -253,16 +276,21 @@ struct Resolver {
...
 
 	template <class... T>
-	static constexpr auto&& extract(T... t)
+	static constexpr auto&& extract_impl(T... t)
 	{
 		if constexpr (TYPE == PIT_STATIC_L0) {
 			return std::get<POS>(std::tie(t...)).get();
 		} else if constexpr (is_path_item_static(PI)) {
-			return tnt::get<POS>(prev(t...));
+			return tnt::get<POS>(parent_container(t...));
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:293]: Nit: looks like this hunk belongs to first patch.

File: src/mpp/Dec.hpp [hunk 3]
```diff
@@ -645,8 +705,7 @@ struct JumpsBuilder {
 		if constexpr (path_item_type(LAST) == PIT_DYN_SKIP) {
 			return getFamiliesByRules<all_rules_t>();
 		} else if constexpr (path_item_type(LAST) == PIT_DYN_KEY) {
-			using R = decltype(path_resolve(PATH{},
-							std::declval<T>()...));
+			using R = decltype(path_resolve(PATH{}, std::declval<T>()...));
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:708]: Nit: look's like an unnecessary hunk.

File: src/mpp/Dec.hpp [hunk 4]
```diff
@@ -659,8 +718,7 @@ struct JumpsBuilder {
 				return keyMapKeyFamiliesFlat<DST>(is{});
 			}
 		} else {
-			using R = decltype(path_resolve(PATH{},
-							std::declval<T>()...));
+			using R = decltype(path_resolve(PATH{}, std::declval<T>()...));
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:721]: Nit: look's like an unnecessary hunk.

File: src/mpp/Dec.hpp [hunk 5]
```diff
@@ -857,17 +915,16 @@ constexpr enum path_item_type get_next_arr_item_type()
 template <class ARR, enum path_item_type TYPE>
 constexpr size_t get_next_arr_static_size()
 {
-	if constexpr (TYPE <= PIT_STADYN)
+	if constexpr (TYPE > PIT_BAD && TYPE <= PIT_STADYN)
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:918]: Nit: looks like an unrelated change.

File: test/EncDecTest.cpp [hunk 6]
```diff
@@ -1269,11 +1270,18 @@ test_optional()
...
+	fail_unless(opt_num.value() == 42);
+
 	buf.flush();
 
 	TEST_CASE("containers with numbers");
 	int null_idx = 4;
-	mpp::encode(buf, mpp::as_arr(std::forward_as_tuple(0, 1, 2, 3, nullptr, 5)));
+	mpp::encode(buf, std::make_optional(mpp::as_arr(
+		std::forward_as_tuple(0, std::make_optional(1), 2, 3, std::optional<int>(), 5)
+	)));
```
Comment (CuriousGeorgiy) [inline test/EncDecTest.cpp:1284]: Nit: no sure it's worth it, but I think adding separate test cases would be cleaner than wrapping around existing ones.

Same comment for `objects` test case below.

File: src/mpp/Dec.hpp [hunk 7]
```diff
@@ -936,6 +993,32 @@ bool jump_read(BUF& buf, T... t)
...
+{
+	auto&& dst = unwrap(path_resolve_transparent(PATH{}, t...));
+	return jump_read_impl<FAMILY, SUBRULE, PATH>(buf, dst, t...);
+}
+
+template <compact::Family FAMILY, size_t SUBRULE,
+	  class PATH, class BUF, class... T>
+bool jump_read_optional(BUF& buf, T... t)
+{
+	auto&& optional_dst = unwrap(path_resolve(PATH{}, t...));
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:1011]: Can we add some kind of assert about the type?

File: test/EncDecTest.cpp [hunk 8]
```diff
@@ -1214,11 +1238,206 @@ test_object_codec()
 
 	auto itr = buf.begin();
 	mpp::decode(itr, rd);
-
```
Comment (CuriousGeorgiy) [inline test/EncDecTest.cpp:1217]: Nit: looks like an unnecessary hunk.

File: test/EncDecTest.cpp [hunk 9]
```diff
@@ -1214,11 +1238,206 @@ test_object_codec()
...
 	fail_unless(rd == wr);
+
+	mpp::decode(itr, rds);
+	fail_unless(rds.count(wr) > 0);
```
Comment (CuriousGeorgiy) [inline test/EncDecTest.cpp:1244]: How is this test case related to the patch?

File: src/mpp/Dec.hpp [hunk 10]
```diff
@@ -936,6 +993,32 @@ bool jump_read(BUF& buf, T... t)
...
+	if constexpr (FAMILY == compact::MP_NIL) {
+		[[maybe_unused]] auto val = read_value<FAMILY, SUBRULE>(buf);
+		optional_dst.reset();
+		return decode_next<PATH>(buf, t...);
+	} else {
+		if (!optional_dst.has_value())
+			optional_dst.emplace();
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:1018]: Could you elaborate the purpose of this step?

**Review by CuriousGeorgiy — COMMENTED** [review]
Looks good to me overall. 

The only thing left unclear to me yet is how results are 'sinked' to destinations, and why container types are filled with 'contained' or, in your notion, 'transparent' values, e.g., `int` of `std::optional<int>` (or is that simply an optimization?).

You haven't mentioned this in your commit message, but as far as I understand, the trade-off under question is introducing an intermediate jump, i.e., `jump_read_optional`.

TBH, I didn't review the encoding test cases thoroughly, I think we need to setup a coverage workflow.

File: src/mpp/Dec.hpp [hunk 11]
```diff
@@ -327,18 +374,29 @@ struct Resolver {
...
+/** Resolve path, semi-transparent object are visible. */
+template <size_t... P, class... T>
+constexpr auto&& path_resolve_transparent(tnt::iseq<P...>, T... t)
+{
+	using Res = Resolver<false, sizeof...(P) - 1, P...>;
+	static_assert(Res::expected_arg_count() == sizeof...(T));
+	return Res::get(t...);
+}
+
+/** Resolve path, semi-transparent object are invisible. */
```
Comment (CuriousGeorgiy) [inline src/mpp/Dec.hpp:386]: I guess the comments should be the other way around, i.e., visible for `path_resolve` and invisible for `path_resolve_transparent`.

File: src/mpp/Dec.hpp (hunk 1 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp:86]: The advantage is that it works :)
`detectFamily` returns a `family_sequence` object, which cannot be a template parameter for another `family_sequence`.

File: src/mpp/Dec.hpp (hunk 5 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp:918]: Build fails with compilation error without this change.

File: src/mpp/Dec.hpp (hunk 10 above)
Comment (drewdzzz) [reply src/mpp/Dec.hpp:1018]: Later, we want to use underlying value directly. For this purpose, we should emplace `optional` is it's empty.

**Comment by drewdzzz** [comment]
Another implementation of optional has been merged just now.
